packages/@aws-cdk/aws-cloudfront/lib/private/cache-behavior.ts

```diff
@@ -52,6 +52,11 @@ export class CacheBehavior {
           if (edgeLambda.functionVersion.version === '$LATEST') {
             throw new Error('$LATEST function version cannot be used for Lambda@Edge');
           }
+
+          if (edgeLambda.functionVersion.lambda.removeEnvironment()) {
+            edgeLambda.functionVersion.lambda.node.addWarning(`Removed environment variables from function ${edgeLambda.functionVersion.lambda.node.path} because Lambda@Edge does not support environment variables`);
+          }
```

Comment:
This is going to surprise customers who are not aware of this limitation and expect their environment variables to show up.

How about removing environment variables that the CDK inserts, and throwing an error when the customer has environment variables configured?

Reply:
Also (FYI), the `addWarning` API was just deprecated yesterday (#9563) as part of prep work for CDK 2.0 [(Design)](https://github.com/aws/aws-cdk-rfcs/blob/master/text/0192-remove-constructs-compat.md#09-logging).

In the future, you'd use:
```suggestion
          if (edgeLambda.functionVersion.lambda.removeEnvironment()) {
            Annotations.of(edgeLambda.functionVersion.lambda).addWarning(`Removed environment variables from function ${edgeLambda.functionVersion.lambda.node.path} because Lambda@Edge does not support environment variables`);
          }
```

Reply:
@nija-at 
This means that we need to identify env vars inserted by CDK? You would like to add a special method (`addCdkEnvironment()`) and prop (`cdkEnvironment`) to identify those? Doesn't it require "developer discipline" inside the CDK?

Also consider the following use case outside the CDK: you have a construct that creates your special Lambda with a defined lists of env vars (whether used in the runtime code or not, like `STAGE` for example) and other good defaults. You want to use it both at edge and not.

@njlynch 
saw #9563 but unfortunately `Annotations` has not been exported in `core/lib/index.ts` in this PR. Will be done when all `node.addXxx()` methods will be updated in the constructs library I presume.

Reply:
That's fair @jogold.

In your example, if I understand the lambda@edge limitation correctly, the special lambda function in the construct may not work at edge if it relies on certain env vars being present.

In this case, does it make sense instead to have a property `allowInEdgeEnv` or `edgeEnvOptimized` or something similar (`false` by default), that strips out env vars when used in edge environment, and errors out otherwise?

This can be used to encode similar behaviour for other [lambda@edge limitations](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-requirements-limits.html#lambda-requirements-lambda-function-configuration).

Reply:
@nija-at how about the following?

* add a `get edgeArn()` getter on `IVersion`: in there we can check that the version is "compatible" for Lambda@Edge and throw otherwise, let's start with env vars only in this PR (+ the check [that it's not $LATEST](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-cloudfront/lib/private/cache-behavior.ts#L52-L54))
* call this `get edgeArn()` in `aws-cloudfront` instead of doing [`lambdaFunctionArn: edgeLambda.functionVersion.functionArn`](https://github.com/aws/aws-cdk/blob/a772fe84784e62843ef724a9158fc8cda848c5c9/packages/%40aws-cdk/aws-cloudfront/lib/private/cache-behavior.ts#L56) => `lambdaFunctionArn: edgeLambda.functionVersion.edgeArn`.
* add a third parameter `allowEdgeRemoval` to `addEnvironment()` so that we can loop over the env vars in `get edgeArn()`, remove the env vars that we're allowed to and throw if anything remains (with a clear message saying that `addEnvironment()` can be used to overcome it). `allowEdgeRemoval` will be set to `true` in the [call to `addEnvironment()` in `aws-lambda-nodejs`](https://github.com/aws/aws-cdk/blob/a772fe84784e62843ef724a9158fc8cda848c5c9/packages/%40aws-cdk/aws-lambda-nodejs/lib/function.ts#L88)

Reply:
This approach seems reasonable to me. 

Minor tweaks - make the third parameter to `addEnvironment()` - `EnvironmentOptions` - which holds the property `removeInEdge` (or something like that).